lib/pos_blockchain.go

```diff
@@ -26,20 +47,129 @@ import (
+//  2. Do nothing if the header is an orphan.
+//  3. Validate the header and verify that its parent is also valid.
+//  4. Add the block header to the block index with status
 //     StatusHeaderValidated or StatusHeaderValidateFailed.
-//  3. Perform the orphan check on the block header.
-//  4. If the block is an orphan, or its view is less than the current header chain's tip,
-//     then we exit early.
-//  5. If it is not an orphan, and has a higher view than the current header chain, then
-//     we re-org the header chain so that the incoming header is the new tip.
-func (bc *Blockchain) processHeaderPoS(blockHeader *MsgDeSoHeader, headerHash *BlockHash) (
-	_isMainChain bool,
-	_isOrphan bool,
-	_err error,
+//  5. Exit early if the's view is less than the current header chain's tip.
+//  6. Reorg the best header chain if the header's view is higher than the current tip.
+func (bc *Blockchain) processHeaderPoS(header *MsgDeSoHeader, headerHash *BlockHash) (
+	_isMainChain bool, _isOrphan bool, _err error,
 ) {
-	return false, false, fmt.Errorf("processHeaderPoS: Not implemented")
+	// Validate the header and index it in the block index.
+	blockNode, isOrphan, err := bc.validateAndIndexHeaderPoS(header, headerHash)
+	if err != nil {
+		return false, false, errors.Wrapf(err, "processHeaderPoS: Problem validating and indexing header: ")
+	}
+
+	// Exit early if the header is an orphan.
+	if isOrphan {
+		return false, true, nil
+	}
+
+	// Exit early if the header's view is less than the current header chain's tip. The header is not
+	// the new tip for the best header chain.
+	currentTip := bc.headerTip()
+	if header.ProposedInView <= currentTip.Header.ProposedInView {
+		return false, false, nil
+	}
+
+	// The header is not an orphan and has a higher view than the current tip. We reorg the header chain
+	// and apply the incoming header is the new tip.
+	_, blocksToDetach, blocksToAttach := GetReorgBlocks(currentTip, blockNode)
+	bc.bestHeaderChain, bc.bestHeaderChainMap = updateBestChainInMemory(
+		bc.bestHeaderChain,
+		bc.bestHeaderChainMap,
+		blocksToDetach,
+		blocksToAttach,
+	)
+
+	// Success. The header is at the tip of the best header chain.
+	return true, false, nil
+}
+
+func (bc *Blockchain) validateAndIndexHeaderPoS(header *MsgDeSoHeader, headerHash *BlockHash) (
+	_headerBlockNode *BlockNode, _isOrphan bool, _err error,
+) {
+	// Look up the header in the block index to check if it has already been validated and indexed.
+	blockNode, exists := bc.blockIndexByHash[*headerHash]
+
+	// ------------------------------------ Base Cases ----------------------------------- //
+
+	// The header is already validated. Exit early.
+	if exists && blockNode.IsHeaderValidated() {
+		return blockNode, false, nil
+	}
+
+	// The header has already failed validations. Exit early.
+	if exists && blockNode.IsHeaderValidateFailed() {
+		return nil, false, errors.New("validateAndIndexHeaderPoS: Header already failed validation")
```

Comment:
Should we return the block node here? It doesn't really matter based on how it's currently being used, but this is how `validateAndIndexBlockPoS` behaves.

Reply:
There's no need, IMO. Currently, the block node is returned only if the header is validated and can be applied as the tip. It's a simpler spec than if we returned it in some cases where it's also invalid (but not all cases)